tmc2130: Fix wrong values report by M913 when Hold > Run and more bugs

gudnimg opened the pull request:
This PR fixes 3 bugs:

1. M913 doesn't report correct current values if Hold > Run and Hold current was truncated to match Run current
2. Hold current isn't updated correctly if the vSense bit changes due to Run current being increased or decreased (the case when Run decreases was not handled at all)
3. Fixed wrong order of register writes in `tmc2130_init` which were unintentionally changed in PFW-1392. CHOPCONF and TPOWERDOWN were written last (3.13.1) instead of first (3.13.0 and before).

----

The firmware will ensure that the Hold current can never exceed the Run current. In this scenario we must update the global current array so that M913 reflects the actual register settings.

Added a echo to serial when this truncation happens.

Steps to reproduce (MK3S+):

1. Enable `TMC2130_SERVICE_CODES_M910_M918`
2. Boot up printer and send `M913` to get default current values
```
tmc2130_print_currents()
	H	R
X	16	16
Y	20	20
Z	17	17
E	30	30
ok
```
4. Send `M911 Z35 E35`
5. Send M913 again and compare
```
tmc2130_print_currents()
	H	R
X	16	16
Y	20	20
Z	35	17
E	35	30
ok
```
7. Observe issue

✅ **Expected behavior:** M913 should show same values in steps 5) and 2)
:x: **Actual behavior:** M913 shows different values

Kudos to @ojrik for finding this issue 😄 

Change in memory:
Flash: +136 bytes
SRAM: 0 bytes

gudnimg committed:
tmc2130: Fix wrong values report by M913 when Hold > Run

The firmware will ensure that the Hold current can never
exceed the Run current. In this scenario we must update
the global current array so that M913 reflects the register settings.

Added a echo to serial when this truncation happens

Change in memory:
Flash: +54 bytes
SRAM: 0 bytes

gudnimg commented:
I'm open to suggestions to shorten that log message :)

gudnimg commented:
There is another separate issue. If vSense bit is 0 and Run current is changed such that vSense = 1 again, then Hold current is not changed to reflect the change in vSense.

gudnimg committed:
tmc2130: Fix bug with vSense bit and current values

If vSense changes at runtime due to Run current
being changed. Then we must always shift the Hold current
correctly. Whether the vSense is changing 1 -> 0 or 0 ->1

Change in memory (with TMC2130_SERVICE_CODES_M910_M918):
Flash: +76 bytes
SRAM: 0 bytes

gudnimg commented:
@ojrik I just added another commit which should resolve issues with changing the currents via M911 and M912. Does this improve things on your 12V printer or are you seeing issues as soon as the printer boots up?

This fix may not be final but it seems to work on my end

ojrik commented:
> 

Greetings. I am currently printing parts for the mk3s 12v. I'll try it later.

gudnimg commented:
Just to satisfy my curiosity I logged all the data being written to the motor drivers at boot-up to be 100% sure the settings are still the same.


3.13.0
=====

```
start
echo: 3.13.0-6875
SpoolJoin is Off
echo: Last Updated: 1970-01-01 00:00:00 | Author: (none, default config)
echo: Free Memory: 2514  PlannerBufferBytes: 1760
X axis, address 0xec, value 0x000000d3       ; CHOPCONF
X axis, address 0x90, value 0x00001010       ; IHOLD_IRUN 
Y axis, address 0xec, value 0x000000d3       ; CHOPCONF
Y axis, address 0x90, value 0x00001414       ; IHOLD_IRUN 
Z axis, address 0xec, value 0x000000d3       ; CHOPCONF
Z axis, address 0x90, value 0x00001111       ; IHOLD_IRUN
E axis, address 0xec, value 0x000000d3       ; CHOPCONF
E axis, address 0x90, value 0x00001e1e       ; IHOLD_IRUN
echo:Stored settings retrieved
adc_init
Hotend fan type: NOCTUA
CrashDetect ENABLED!
X axis, address 0xec, value 0x000000d3       ; CHOPCONF
X axis, address 0x90, value 0x00001010       ; IHOLD_IRUN
X axis, address 0x91, value 0x00000000       ; TPOWERDOWN
X axis, address 0xed, value 0x00000000
X axis, address 0x94, value 0x000001ae
X axis, address 0x80, value 0x00000180
X axis, address 0xf0, value 0x000002e6
X axis, address 0x93, value 0x00000000
Y axis, address 0xec, value 0x000000d3       ; CHOPCONF
Y axis, address 0x90, value 0x00001414       ; IHOLD_IRUN
Y axis, address 0x91, value 0x00000000       ; TPOWERDOWN
Y axis, address 0xed, value 0x00000000
Y axis, address 0x94, value 0x000001ae
Y axis, address 0x80, value 0x00000180
Y axis, address 0xf0, value 0x000002eb
Y axis, address 0x93, value 0x00000000
Z axis, address 0xec, value 0x000000d3       ; CHOPCONF
Z axis, address 0x90, value 0x00001111       ; IHOLD_IRUN
Z axis, address 0x91, value 0x00000000       ; TPOWERDOWN
Z axis, address 0xed, value 0x00000000
Z axis, address 0x94, value 0x000001f4
Z axis, address 0x80, value 0x00000184
Z axis, address 0xf0, value 0x000004c8
Z axis, address 0x93, value 0x0000fff0
E axis, address 0xec, value 0x000000d3       ; CHOPCONF
E axis, address 0x90, value 0x00001e1e       ; IHOLD_IRUN
E axis, address 0x91, value 0x00000000       ; TPOWERDOWN
E axis, address 0x80, value 0x00000180
X axis, address 0xe9, value 0x00000000
X axis, address 0xe0, value 0x0000b556
X axis, address 0xe1, value 0x000054aa
X axis, address 0xe2, value 0x00002929
X axis, address 0xe3, value 0x00004222
X axis, address 0xe4, value 0x00000000
X axis, address 0xe5, value 0x0000777d
X axis, address 0xe6, value 0x00005556
X axis, address 0xe7, value 0x00004222
X axis, address 0xe8, value 0x00000259
Y axis, address 0xe9, value 0x00000000
Y axis, address 0xe0, value 0x0000b556
Y axis, address 0xe1, value 0x000054aa
Y axis, address 0xe2, value 0x00002929
Y axis, address 0xe3, value 0x00004222
Y axis, address 0xe4, value 0x00000000
Y axis, address 0xe5, value 0x0000777d
Y axis, address 0xe6, value 0x00005556
Y axis, address 0xe7, value 0x00004222
Y axis, address 0xe8, value 0x00000259
Z axis, address 0xe9, value 0x00000000
Z axis, address 0xe0, value 0x0000b556
Z axis, address 0xe1, value 0x000054aa
Z axis, address 0xe2, value 0x00002929
Z axis, address 0xe3, value 0x00004222
Z axis, address 0xe4, value 0x00000000
Z axis, address 0xe5, value 0x0000777d
Z axis, address 0xe6, value 0x00005556
Z axis, address 0xe7, value 0x00004222
Z axis, address 0xe8, value 0x00000259
E axis, address 0xe9, value 0x00000000
E axis, address 0xe0, value 0x0000b556
E axis, address 0xe1, value 0x000054aa
E axis, address 0xe2, value 0x00002929
E axis, address 0xe3, value 0x00004222
E axis, address 0xe4, value 0x00000000
E axis, address 0xe5, value 0x0000777d
E axis, address 0xe6, value 0x00005556
E axis, address 0xe7, value 0x00004222
E axis, address 0xe8, value 0x00000259
X axis, address 0xec, value 0x000000d3       ; CHOPCONF
X axis, address 0x90, value 0x00001010       ; IHOLD_IRUN
X axis, address 0x91, value 0x00000000       ; TPOWERDOWN
X axis, address 0xed, value 0x00000000
X axis, address 0x94, value 0x000001ae
X axis, address 0x80, value 0x00000180
X axis, address 0xf0, value 0x000002e6
X axis, address 0x93, value 0x00000000
Y axis, address 0xec, value 0x000000d3       ; CHOPCONF
Y axis, address 0x90, value 0x00001414       ; IHOLD_IRUN
Y axis, address 0x91, value 0x00000000       ; TPOWERDOWN
Y axis, address 0xed, value 0x00000000
Y axis, address 0x94, value 0x000001ae
Y axis, address 0x80, value 0x00000180
Y axis, address 0xf0, value 0x000002eb
Y axis, address 0x93, value 0x00000000
Z axis, address 0xec, value 0x000000d3       ; CHOPCONF
Z axis, address 0x90, value 0x00001111       ; IHOLD_IRUN
Z axis, address 0x91, value 0x00000000       ; TPOWERDOWN
Z axis, address 0xed, value 0x00000000
Z axis, address 0x94, value 0x000001f4
Z axis, address 0x80, value 0x00000184
Z axis, address 0xf0, value 0x000004c8
Z axis, address 0x93, value 0x0000fff0
E axis, address 0xec, value 0x000000d3       ; CHOPCONF
E axis, address 0x90, value 0x00001e1e       ; IHOLD_IRUN
E axis, address 0x91, value 0x00000000       ; TPOWERDOWN
E axis, address 0x80, value 0x00000180
X axis, address 0xe9, value 0x00000000
X axis, address 0xe0, value 0x0000b556
X axis, address 0xe1, value 0x000054aa
X axis, address 0xe2, value 0x00002929
X axis, address 0xe3, value 0x00004222
X axis, address 0xe4, value 0x00000000
X axis, address 0xe5, value 0x0000777d
X axis, address 0xe6, value 0x00005556
X axis, address 0xe7, value 0x00004222
X axis, address 0xe8, value 0x00000259
Y axis, address 0xe9, value 0x00000000
Y axis, address 0xe0, value 0x0000b556
Y axis, address 0xe1, value 0x000054aa
Y axis, address 0xe2, value 0x00002929
Y axis, address 0xe3, value 0x00004222
Y axis, address 0xe4, value 0x00000000
Y axis, address 0xe5, value 0x0000777d
Y axis, address 0xe6, value 0x00005556
Y axis, address 0xe7, value 0x00004222
Y axis, address 0xe8, value 0x00000259
Z axis, address 0xe9, value 0x00000000
Z axis, address 0xe0, value 0x0000b556
Z axis, address 0xe1, value 0x000054aa
Z axis, address 0xe2, value 0x00002929
Z axis, address 0xe3, value 0x00004222
Z axis, address 0xe4, value 0x00000000
Z axis, address 0xe5, value 0x0000777d
Z axis, address 0xe6, value 0x00005556
Z axis, address 0xe7, value 0x00004222
Z axis, address 0xe8, value 0x00000259
E axis, address 0xe9, value 0x00000000
E axis, address 0xe0, value 0x0000b556
E axis, address 0xe1, value 0x000054aa
E axis, address 0xe2, value 0x00002929
E axis, address 0xe3, value 0x00004222
E axis, address 0xe4, value 0x00000000
E axis, address 0xe5, value 0x0000777d
E axis, address 0xe6, value 0x00005556
E axis, address 0xe7, value 0x00004222
E axis, address 0xe8, value 0x00000259
```

3.13.1 (this branch)
=====

```
start
echo: 3.13.0-RC1-6853
SpoolJoin is Off
echo: Last Updated: 1970-01-01 00:00:00 | Author: (none, default config)
echo: Free Memory: 2515  PlannerBufferBytes: 1760
X axis, address 0xec, value 0x000000d3       ; CHOPCONF
X axis, address 0x90, value 0x00001010       ; IHOLD_IRUN
Y axis, address 0xec, value 0x000000d3       ; CHOPCONF
Y axis, address 0x90, value 0x00001414       ; IHOLD_IRUN
Z axis, address 0xec, value 0x000000d3       ; CHOPCONF
Z axis, address 0x90, value 0x00001111       ; IHOLD_IRUN
E axis, address 0xec, value 0x000000d3       ; CHOPCONF
E axis, address 0x90, value 0x00001e1e       ; IHOLD_IRUN
echo:Stored settings retrieved
adc_init
Hotend fan type: NOCTUA
CrashDetect ENABLED!
tmc2130_init start
X axis, address 0xed, value 0x00000000
X axis, address 0x94, value 0x000001ae
X axis, address 0x80, value 0x00000180
X axis, address 0xf0, value 0x000002e6
X axis, address 0x93, value 0x00000000
X axis, address 0xec, value 0x000000d3       ; CHOPCONF
X axis, address 0x90, value 0x00001010       ; IHOLD_IRUN
X axis, address 0x91, value 0x00000000       ; TPOWERDOWN
Y axis, address 0xed, value 0x00000000
Y axis, address 0x94, value 0x000001ae
Y axis, address 0x80, value 0x00000180
Y axis, address 0xf0, value 0x000002eb
Y axis, address 0x93, value 0x00000000
Y axis, address 0xec, value 0x000000d3       ; CHOPCONF
Y axis, address 0x90, value 0x00001414       ; IHOLD_IRUN
Y axis, address 0x91, value 0x00000000       ; TPOWERDOWN
Z axis, address 0xed, value 0x00000000
Z axis, address 0x94, value 0x000001f4
Z axis, address 0x80, value 0x00000184
Z axis, address 0xf0, value 0x000004c8
Z axis, address 0x93, value 0x0000fff0
Z axis, address 0xec, value 0x000000d3       ; CHOPCONF
Z axis, address 0x90, value 0x00001111       ; IHOLD_IRUN
Z axis, address 0x91, value 0x00000000       ; TPOWERDOWN
E axis, address 0xec, value 0x000000d3       ; CHOPCONF
E axis, address 0x90, value 0x00001e1e       ; IHOLD_IRUN
E axis, address 0x91, value 0x00000000       ; TPOWERDOWN
E axis, address 0x80, value 0x00000180
X axis, address 0xe9, value 0x00000000
X axis, address 0xe0, value 0x0000b556
X axis, address 0xe1, value 0x000054aa
X axis, address 0xe2, value 0x00002929
X axis, address 0xe3, value 0x00004222
X axis, address 0xe4, value 0x00000000
X axis, address 0xe5, value 0x0000777d
X axis, address 0xe6, value 0x00005556
X axis, address 0xe7, value 0x00004222
X axis, address 0xe8, value 0x00000259
Y axis, address 0xe9, value 0x00000000
Y axis, address 0xe0, value 0x0000b556
Y axis, address 0xe1, value 0x000054aa
Y axis, address 0xe2, value 0x00002929
Y axis, address 0xe3, value 0x00004222
Y axis, address 0xe4, value 0x00000000
Y axis, address 0xe5, value 0x0000777d
Y axis, address 0xe6, value 0x00005556
Y axis, address 0xe7, value 0x00004222
Y axis, address 0xe8, value 0x00000259
Z axis, address 0xe9, value 0x00000000
Z axis, address 0xe0, value 0x0000b556
Z axis, address 0xe1, value 0x000054aa
Z axis, address 0xe2, value 0x00002929
Z axis, address 0xe3, value 0x00004222
Z axis, address 0xe4, value 0x00000000
Z axis, address 0xe5, value 0x0000777d
Z axis, address 0xe6, value 0x00005556
Z axis, address 0xe7, value 0x00004222
Z axis, address 0xe8, value 0x00000259
E axis, address 0xe9, value 0x00000000
E axis, address 0xe0, value 0x0000b556
E axis, address 0xe1, value 0x000054aa
E axis, address 0xe2, value 0x00002929
E axis, address 0xe3, value 0x00004222
E axis, address 0xe4, value 0x00000000
E axis, address 0xe5, value 0x0000777d
E axis, address 0xe6, value 0x00005556
E axis, address 0xe7, value 0x00004222
E axis, address 0xe8, value 0x00000259
tmc2130_init end
tmc2130_init start
X axis, address 0xed, value 0x00000000
X axis, address 0x94, value 0x000001ae
X axis, address 0x80, value 0x00000180
X axis, address 0xf0, value 0x000002e6
X axis, address 0x93, value 0x00000000
X axis, address 0xec, value 0x000000d3       ; CHOPCONF
X axis, address 0x90, value 0x00001010       ; IHOLD_IRUN
X axis, address 0x91, value 0x00000000       ; TPOWERDOWN
Y axis, address 0xed, value 0x00000000
Y axis, address 0x94, value 0x000001ae
Y axis, address 0x80, value 0x00000180
Y axis, address 0xf0, value 0x000002eb
Y axis, address 0x93, value 0x00000000
Y axis, address 0xec, value 0x000000d3       ; CHOPCONF
Y axis, address 0x90, value 0x00001414       ; IHOLD_IRUN
Y axis, address 0x91, value 0x00000000       ; TPOWERDOWN
Z axis, address 0xed, value 0x00000000
Z axis, address 0x94, value 0x000001f4
Z axis, address 0x80, value 0x00000184
Z axis, address 0xf0, value 0x000004c8
Z axis, address 0x93, value 0x0000fff0
Z axis, address 0xec, value 0x000000d3       ; CHOPCONF
Z axis, address 0x90, value 0x00001111       ; IHOLD_IRUN
Z axis, address 0x91, value 0x00000000       ; TPOWERDOWN
E axis, address 0xec, value 0x000000d3       ; CHOPCONF
E axis, address 0x90, value 0x00001e1e       ; IHOLD_IRUN
E axis, address 0x91, value 0x00000000       ; TPOWERDOWN
E axis, address 0x80, value 0x00000180
X axis, address 0xe9, value 0x00000000
X axis, address 0xe0, value 0x0000b556
X axis, address 0xe1, value 0x000054aa
X axis, address 0xe2, value 0x00002929
X axis, address 0xe3, value 0x00004222
X axis, address 0xe4, value 0x00000000
X axis, address 0xe5, value 0x0000777d
X axis, address 0xe6, value 0x00005556
X axis, address 0xe7, value 0x00004222
X axis, address 0xe8, value 0x00000259
Y axis, address 0xe9, value 0x00000000
Y axis, address 0xe0, value 0x0000b556
Y axis, address 0xe1, value 0x000054aa
Y axis, address 0xe2, value 0x00002929
Y axis, address 0xe3, value 0x00004222
Y axis, address 0xe4, value 0x00000000
Y axis, address 0xe5, value 0x0000777d
Y axis, address 0xe6, value 0x00005556
Y axis, address 0xe7, value 0x00004222
Y axis, address 0xe8, value 0x00000259
Z axis, address 0xe9, value 0x00000000
Z axis, address 0xe0, value 0x0000b556
Z axis, address 0xe1, value 0x000054aa
Z axis, address 0xe2, value 0x00002929
Z axis, address 0xe3, value 0x00004222
Z axis, address 0xe4, value 0x00000000
Z axis, address 0xe5, value 0x0000777d
Z axis, address 0xe6, value 0x00005556
Z axis, address 0xe7, value 0x00004222
Z axis, address 0xe8, value 0x00000259
E axis, address 0xe9, value 0x00000000
E axis, address 0xe0, value 0x0000b556
E axis, address 0xe1, value 0x000054aa
E axis, address 0xe2, value 0x00002929
E axis, address 0xe3, value 0x00004222
E axis, address 0xe4, value 0x00000000
E axis, address 0xe5, value 0x0000777d
E axis, address 0xe6, value 0x00005556
E axis, address 0xe7, value 0x00004222
E axis, address 0xe8, value 0x00000259
tmc2130_init end
```

gudnimg commented:
Looking at my logs above I don't see any difference in the TMC2130 settings in 3.13.0 compared to this branch 🤔 at least on MK3S+

One difference I spot is different order in which registers are written. But the SPI content is the same so I'm not sure if this may be something.

gudnimg committed:
tmc2130: write chopper config and powerdown setting first

I noticed this is how the order is in 3.13.0 and before.
I want to keep it exactly the same.

This somehow saves 2 bytes of flash. Probably compiler magic.

Change in memory:
Flash: -2 bytes
SRAM: 0 bytes

gudnimg commented:
Added another commit which changes the order of the TMC2130 registers initialization to match 3.13.0, it was definitely not intentional to change this.

3.13.0:
```
X axis, address 0xec, value 0x000000d3       ; CHOPCONF
X axis, address 0x90, value 0x00001010       ; IHOLD_IRUN
X axis, address 0x91, value 0x00000000       ; TPOWERDOWN
X axis, address 0xed, value 0x00000000
X axis, address 0x94, value 0x000001ae
X axis, address 0x80, value 0x00000180
X axis, address 0xf0, value 0x000002e6
X axis, address 0x93, value 0x00000000
```

3.13.1 after [PFW-1392](https://github.com/prusa3d/Prusa-Firmware/pull/4191) (fixed now such that it is the same as 3.13.0)

```
X axis, address 0xed, value 0x00000000
X axis, address 0x94, value 0x000001ae
X axis, address 0x80, value 0x00000180
X axis, address 0xf0, value 0x000002e6
X axis, address 0x93, value 0x00000000
X axis, address 0xec, value 0x000000d3       ; CHOPCONF
X axis, address 0x90, value 0x00001010       ; IHOLD_IRUN
X axis, address 0x91, value 0x00000000       ; TPOWERDOWN
```

ojrik commented:
Greetings. I don't want to start the MK3S+ 12V for a long time. It already works for me, so far I made a test print with steps of 0.15mm. After 3 layers, I increased the speed to 150%, then to 400%. I watched while the M913 was printing, the machine adjusted the H value. The printing went without a problem. I also adjusted the values ​​in MK3S define:

#define TMC2130_CURRENTS_H {15, 15, 18, 21}
#define TMC2130_CURRENTS_R {30, 30, 35, 42}
#define TMC2130_CURRENTS_R_HOME {20, 20, 35, 25}


![20230731_080028](https://github.com/prusa3d/Prusa-Firmware/assets/29524917/ddeec665-541d-4047-8c28-d5c5b79a10dd)
Test
